### ISSUE TITLE
Merge `persist_sink` implementations

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#18076]
*Part of the multipurpose clusters epic (#17413)*.

We currently have two implementations of the `persist_sink`: one in storage and one in compute.

Before allowing multipurpose clusters, the compute team has requested that we merge these implementations, to reduce the ongoing maintenance burden. This is a hard blocker for #17659.

There is quite a bit of historical context about why these implementations are separate in this design document: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20230210_storage_persist_sink_redux.md#conclusion-and-alternatives.

cc @antiguru @guswynn

**Comment by aljoscha** [comment]
I believe we had "i.e., merging the persist sink or finding a sound argument why we shouldn't". It might still turn out that the requirements we have for the implementation are just too different.

**Comment by antiguru** [comment]
Yes, I support what @aljoscha said. Compute is fine with two different implementations if they would not share a significant portion of their implementation internally. However, we need to determine whether this would be true!

**Comment by benesch** [comment]
Gotcha, thanks! I've updated the issue description accordingly.
